File: TopCoder/src/main/java/com/leetcode/array/leetcode/FindFirstAndLastElementInSortedArray.java
package com.leetcode.array.leetcode;

//https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
public class FindFirstAndLastElementInSortedArray {

    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0) {
            return new int[]{-1, -1};
        }
        int firstIndex = findFirstIndex(nums, target, 0, nums.length-1);
        if (firstIndex == -1) {
            return new int[]{-1, -1};
        }
        int secondndex = findSecondIndex(nums, target, firstIndex, nums.length-1);

        return new int[]{firstIndex, secondndex};
    }

    private int findSecondIndex(int[] nums, int target, int low, int high) {
        if (low > high) {
            return -1;
        }

        int mid = low + (high - low)/2;

        if (nums[mid] == target && (mid + 1 >= nums.length || nums[mid+1] != target)) {
            return mid;
        } else {
            if (nums[mid] <= target) {
                return findSecondIndex(nums, target, mid + 1, high);
            } else {
                return findSecondIndex(nums, target, low, mid - 1);
            }
        }

    }

    private int findFirstIndex(int[] nums, int target, int low, int high) {
        if (low > high) {
            return -1;
        }

        int mid = low + (high - low)/2;

        if (nums[mid] == target && (mid - 1 < 0 || nums[mid-1] != target)) {
            return mid;
        } else {
            if (nums[mid] < target) {
                return findFirstIndex(nums, target, mid+1, high);
            } else {
                return findFirstIndex(nums, target, low, mid-1);
            }
        }
    }

    public static void main(String[] args) {
        int arr[] = {2,2};
        int[] result = new FindFirstAndLastElementInSortedArray().searchRange(arr, 2);
        for (int i : result) {
            System.out.println(i);
        }
    }
}

File: TopCoder/python/FindFirstAndLastElementInSortedArray.py
from typing import List

class Solution:

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums or len(nums) == 0:
            return [-1, -1]

        first_index = self.find_first_index(nums, 0, len(nums)-1, target)
        if first_index == -1:
            return [-1, -1]

        second_index = self.find_second_index(nums, first_index, len(nums)-1, target)
        return [first_index, second_index]

    def find_first_index(self, nums, low, high, target):
        if low > high:
            return -1

        mid = int(low + (high - low)/2)

        if nums[mid] == target and (mid - 1 < 0 or nums[mid - 1] != target):
            return mid

        if nums[mid] < target:
            return self.find_first_index(nums, mid + 1, high, target)
        else:
            return self.find_first_index(nums, low, mid - 1, target)

    def find_second_index(self, nums, low, high, target):
        if low > high:
            return -1

        mid = int(low + (high - low)/2)
        if nums[mid] == target and (mid + 1 >= len(nums) or nums[mid + 1] != target):
            return mid

        if nums[mid] <= target:
            return self.find_second_index(nums, mid + 1, high, target)
        else:
            return self.find_second_index(nums, low, mid - 1, target)

nums = [2,2]
target = 2
sol = Solution()
result = sol.searchRange(nums, target)
for i, val in enumerate(result):
    print(val)